Fix cgroups v1 for the PREEMPT_RT kernel

rouming opened the pull request:
This PR aims the goal to fix the cgroups v1 for the PREEMPT_RT kernel

    mm/memcg: enable cgroups v1 for the PREEMPT_RT case

    Some of the cgroups v1 knobs were disabled for the PREEMPT_RT
    case in this commit:
    
    2343e88d238f ("mm/memcg: disable threshold event handlers on PREEMPT_RT")
    
    EVE highly depends on the cgroups v1 and unfortunately there is no
    easy way to quickly jump to the cgroups v2.
    
    This patch enables all sysfs entries for the cgroups v1 for the
    PREEMPT_RT case by replacing `local_irq_{disable|save}` with
    safe `local_lock_{irq|irqsave}` alternative, which covers safely
    both kernels: non-PREEMPT_RT and PREEMPT_RT.

Also enables enables kernel debug symbols and NO_HZ_FULL support.

CC: @eriknordmark

rouming committed:
mm/memcg: enable cgroups v1 for the PREEMPT_RT case

Some of the cgroups v1 knobs were disabled for the PREEMPT_RT
case in this commit:

2343e88d238f ("mm/memcg: disable threshold event handlers on PREEMPT_RT")

EVE highly depends on the cgroups v1 and unfortunately there is no
easy way to quickly jump to the cgroups v2.

This patch enables all sysfs entries for the cgroups v1 for the
PREEMPT_RT case by replacing `local_irq_{disable|save}` with
safe `local_lock_{irq|irqsave}` alternative, which covers safely
both kernels: non-PREEMPT_RT and PREEMPT_RT.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
eve_defconfig: enable kernel debug symbols

Very handy to have those around.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rucoder commented on arch/x86/configs/eve_defconfig:
```diff
@@ -6191,11 +6191,18 @@ CONFIG_DEBUG_MISC=y
 CONFIG_AS_HAS_NON_CONST_LEB128=y
-CONFIG_DEBUG_INFO_NONE=y
-# CONFIG_DEBUG_INFO_DWARF_TOOLCHAIN_DEFAULT is not set
+# CONFIG_DEBUG_INFO_NONE is not set
```
I would do one more commit to convert eve_defconfig into defconfig format: 
```
make eve_defconfig
make savedefconfig
cp defconfig arch/x86/configs/eve_defconfig
```

rouming commented:
@eriknordmark, @rucoder ping. 
suggestions? comments? or let's merge that. I'm testing that on the real hpe-m750, all good, eve works from the box (cgroups v1)

eriknordmark commented:
I don't know which use cases (hardware plus applications) use RS232 and RS483, so it might make sense to ask around. If people run with serial chips with very limited buffering and high baud rate the risk is loosing some characters if this slows down the max throughput. But if the larger limit is merely for CPU efficiency, then we can decrease it to get better RT behavior.

What do other RT kernels/users do?

rouming commented:
@eriknordmark the whole discussion happened here, in this (flattened) thread:
https://lore.kernel.org/lkml/nycvar.YEU.7.76.2004231017470.4730@gjva.wvxbf.pm/t/#u

There is no proposed solution (the problem happens on qemu only as far as I concern), but reducing the PASS_LIMIT was discussed here, the value was set to 512 (higher) though:
https://lore.kernel.org/lkml/20200423162004.ogbhgrkiwaphhh2r@linutronix.de/

the following message: 
```
   There is no need to loop for longer. The message of too much work was
   removed in commit
     9d7c249a1ef9b ("serial: 8250: drop the printk from serial8250_interrupt()")
```

As far as I understand the code, there won't be bytes loss observed if the PASS_LIMIT is reduced. The value affects only the "time" we spend in the thread, which reads bytes from serial. The main problem is that on RT the irqthread likely will be preempted by other thread, so the execution of the serial handling can take much longer, which causes unhandled interrupt, which comes while the receiving thread handles the whole PASS_LIMIT.

Reducing the PASS_LIMIT allocates much smaller time slice for the serial processing, which gives interrupt to be handled in time.

So my dilemma is that we can keep as-is, making the qemu case (virtualized EVE) non-functional (the first `logread` invocation causes the described kernel backtrace and then irq4 interrupts are disabled which causes console hang), or I reduce the PASS_LIMIT, which fixes the qemu case, but introduces much smaller queue.

My current assumption is that this kernel branch is rather testing (prove of concept branch), so closing showstopper problems (non-function serial, non-functional cgroups v1) has higher priority.

rucoder commented:
> @eriknordmark, @rucoder ping. suggestions? comments? or let's merge that. I'm testing that on the real hpe-m750, all good, eve works from the box (cgroups v1)

I submitted my comment 2 weeks ago but forgot to commit the review

rucoder reviewed (commented):
The caption says 'reduce PASS_LIMIT to 64' but it is set to 16 in code

rouming commented:
> The caption says 'reduce PASS_LIMIT to 64' but it is set to 16 in code

Drifted away while testing :) Will change. Thanks.

rouming committed:
x86/configs/eve_defconfig: make defconfig

Simplified config

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming committed:
x86/configs/eve_defconfig: enable NO_HZ_FULL

Disables the tick whenever possible if set as `nohz_full=cpuset`
kernel option. Good combination with CPU isolation for the
userspace QEMU tasks.

Signed-off-by: Roman Penyaev <r.peniaev@gmail.com>

rouming commented:
@eriknordmark @rucoder 

Difference to the previous version:
1. So in order to reduce any confusion with the serial fix for qemu, I removed the patch (anyway it is needed only for qemu case and running preempt_rt in qemu does not make a lot of sense now).
2. Simplified config by doing the suggested 
```
make eve_defconfig
make savedefconfig
cp defconfig arch/x86/configs/eve_defconfig
```

rouming commented:
@eriknordmark @rucoder ping

eriknordmark reviewed (commented):
I don't see anything wrong with the PR but I'm not an expert on the kernel nor on how we do the kconfig in the eve-kernel repo. So @rucoder can you take a look?
